kong/runloop/handler.lua

```diff
@@ -1427,6 +1427,19 @@ return {
         end
       end
 
+      -- if this is the last try and it failed, save its state to correctly log it
+      local status = ngx.status
+      if status > 499 and ctx.balancer_data then
```

Comment:
Is this correct? any 500 now classifies as a failure.

The "tries" are bound to the semantics of the nginx retry mechanism and settings. Since this runs in the `header_filter` phase it will apply to every 500 (eg. a plugin that fails in the `access` phase, that would still cause the `header_filter` phase to be executed). It would record the failure, but it has completely different semantics, since effectively there never was a TCP connect attempt to the upstream since we errorred out before.

The reason the last try wasn't recorded is because the `balancer` phase is not called again. The only way to do this proper would be to replicate the retry mechanism in Kong, and no longer rely on Nginx to do that.

Reply:
(sorry for the late comment, the PR close message in Slack caught my eye)